L1Trigger/TrackFindingTMTT/src/Matrix.cc

```diff
@@ -0,0 +1,372 @@
+    unsigned lCols = this->cols;
+    Matrix result(lRows, rCols, 0.0);
+    for (unsigned i = 0; i < lRows; i++) {
+      for (unsigned j = 0; j < rCols; j++) {
+        for (unsigned k = 0; k < lCols; k++) {
+          result(i, j) += this->mat[i][k] * rhs.mat[k][j];
+        }
+      }
+    }
+    return result;
+  }
+
+  // Cumulative left multiplication of this matrix and another
+  template <typename T>
+  Matrix<T>& Matrix<T>::operator*=(const Matrix<T>& rhs) {
+    Matrix result = (*this) * rhs;
+    (*this) = result;
+    return *this;
+  }
+
+  // Calculate a transpose of this matrix
+  template <typename T>
+  Matrix<T> Matrix<T>::transpose() {
+    Matrix result(cols, rows, 0.0);
+
+    for (unsigned i = 0; i < cols; i++) {
+      for (unsigned j = 0; j < rows; j++) {
+        result(i, j) = this->mat[j][i];
+      }
+    }
+
+    return result;
+  }
+
+  // Recursively calculate the determinant
+  template <typename T>
+  T Matrix<T>::determinant() {
+    T det = 0;
+    if (rows < 1) {
+      std::cerr << "Can't have determinant of matrix with " << rows << " rows." << std::endl;
+    } else if (rows == 1) {
+      det = this->mat[0][0];
+    } else if (rows == 2) {
+      det = this->mat[0][0] * this->mat[1][1] - this->mat[1][0] * this->mat[0][1];
+    } else {
+      for (unsigned i = 0; i < rows; i++) {
+        Matrix m(rows - 1, cols - 1, 0);
+        for (unsigned j = 1; j < rows; j++) {
+          unsigned k = 0;
+          for (unsigned l = 0; l < rows; l++) {
+            if (l == i)
+              continue;
+            m(j - 1, k) = this->mat[j][l];
+            k++;
+          }
+        }
+        T sign;
+        if ((i + 2) % 2 == 0)
+          sign = 1;
+        else
+          sign = -1;
+        det += sign * this->mat[0][i] * m.determinant();
+      }
+    }
+    return (det);
+  }
+
+  template <typename T>
+  Matrix<T> Matrix<T>::cofactor() {
+    Matrix result(rows, cols, 0);
+    Matrix c(rows - 1, cols - 1, 0);
+    if (rows != cols) {
+      std::cerr << "Can only compute cofactor of square matrix." << std::endl;
+      return result;
+    }
+    for (unsigned j = 0; j < rows; j++) {
+      for (unsigned i = 0; i < rows; i++) {
+        unsigned i1 = 0;
+        for (unsigned ii = 0; ii < rows; ii++) {
+          if (ii == i)
+            continue;
+          unsigned j1 = 0;
+          for (unsigned jj = 0; jj < rows; jj++) {
+            if (jj == j)
+              continue;
+            c(i1, j1) = this->mat[ii][jj];
+            j1++;
+          }
+          i1++;
+        }
+
+        T det = c.determinant();
+        T sign;
+        if ((i + j + 2) % 2 == 0)
+          sign = 1;
+        else
+          sign = -1;
+        result(i, j) = sign * det;
+      }
+    }
+    return result;
+  }
+
+  template <typename T>
+  Matrix<T> Matrix<T>::inverse() {
+    if (rows != cols) {
+      std::cerr << "Matrix: cannot invert matrix with " << rows << " rows and " << cols << "cols" << std::endl;
```

Comment:
Not thread safe.

Reply:
Now using ROOT's TMatrix class, so class Matrix has been deleted.